homeassistant/components/bsblan/sensor.py

```diff
@@ -0,0 +1,135 @@
+"""Support for BSB-Lan sensors."""
+
+from __future__ import annotations
+
+from bsblan import BSBLAN, Device, Info, Sensor, StaticState
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_HOST,
+    DEGREE,
+    PERCENTAGE,
+    UnitOfPressure,
+    UnitOfTemperature,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo, format_mac
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from . import HomeAssistantBSBLANData
+from .const import DOMAIN
+from .coordinator import BSBLanUpdateCoordinator
+
+
+def get_unit_of_measurement(unit: str) -> str | None:
+    """Map BSBLAN units to Home Assistant units."""
+    if unit in ("°C", "&deg;C"):
+        return UnitOfTemperature.CELSIUS
+    if unit in ("°F", "&deg;F"):
+        return UnitOfTemperature.FAHRENHEIT
+    if unit == "bar":
+        return UnitOfPressure.BAR
+    if unit == "%":
+        return PERCENTAGE
+    if unit == "°":
+        return DEGREE
+    return None
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up BSBLAN sensor based on a config entry."""
+    data: HomeAssistantBSBLANData = hass.data[DOMAIN][entry.entry_id]
+
+    sensor_data: Sensor = data.coordinator.data["sensor"]
+    entities = []
+
+    for attr_name in dir(sensor_data):
+        if attr_name.startswith("_"):
+            continue  # Skip private attributes
+
+        attr_value = getattr(sensor_data, attr_name)
+        if hasattr(attr_value, "value") and hasattr(attr_value, "unit"):
+            unit = get_unit_of_measurement(attr_value.unit)
+            entities.append(
+                BSBLANSensor(
+                    coordinator=data.coordinator,
+                    client=data.client,
+                    device=data.device,
+                    info=data.info,
+                    static=data.static,
+                    entry=entry,
+                    key=attr_name,
+                    name=attr_name.replace("_", " ").title(),
+                    unit=unit,
+                )
+            )
```

Comment:
Do we know all the possible sensors this can create?

Reply:
It creates for now only 2 sensors. current temp and outside temp. If the sensors are available on the heating system.

Reply:
Why don't we create entity_descriptions and define in the HA code which sensors we expose? The current approach looks extendible, but it will lack the ability for HA to use translations for the entity names. It will also make refactoring in the future a lot harder because nobody knows which entities _could_ be created and if there are models that maybe expose more devices than others.

(To give some context, one of these integrations is `aseko_pool_live`. That integration is now in a state where we don't know if we can refactor it, or if someone with dev experience and the device has the full picture to refactor, because they only have one type of device. So if that is possible I'd rather avoid creating more of these setups so when we read the full picture we can at least see what is happening when)

Reply:
I understand and will have a look.

Reply:
Any good example I should follow?

Reply:
I think either `airgradient` or `mealie` will do. Since you have types, I would recommend to change the returntype of the coordinator to a dataclass
```py
@dataclass
class BSBLanData:
    state: State
    sensor: Sensor
```
This way you can just access `self.coordinator.data.state` with typesafety

Reply:
So the idea is to create a generic sensor for a group of sensors that share behavioral characteristics. By injecting entity_descriptions you can define how this sensor is shown. So it's usually a quick and easy method to create a lot of entities, while still maintaining a list of definitions

Reply:
@joostlek I think it's ready. Just not sure of the coordinator config_entry. I had issues with test and giving the config entry it's own name solved it. But not sure if this is the right way to do it?